File: frost-taproot/src/group.rs
// Mirrors ref/frost/src/lib/group.ts
// High-level share set creation (trusted dealer).

use crate::shares::create_shares;
use crate::types::{DealerShareSet, SecretShareSet};
use crate::vss::{create_share_coeffs, get_share_commits};
use crate::Error;

/// Create a set of secret shares and VSS commitments.
/// Mirrors `create_share_set` in the TS implementation.
pub fn create_share_set(
    threshold: usize,
    share_max: u32,
    secrets: &[[u8; 32]],
) -> Result<SecretShareSet, Error> {
    let coeffs = create_share_coeffs(secrets, threshold);
    let shares = create_shares(&coeffs, share_max)?;
    let vss_commits = get_share_commits(&coeffs);
    Ok(SecretShareSet {
        shares,
        vss_commits,
    })
}

/// Create a dealer share set that also exposes the group public key.
/// The group public key is the first VSS commitment (the constant term's public key).
/// Mirrors `create_dealer_set` in the TS implementation.
pub fn create_dealer_set(
    threshold: usize,
    share_max: u32,
    secrets: &[[u8; 32]],
) -> Result<DealerShareSet, Error> {
    let share_set = create_share_set(threshold, share_max, secrets)?;
    let group_pk = share_set.vss_commits[0];
    Ok(DealerShareSet {
        shares: share_set.shares,
        vss_commits: share_set.vss_commits,
        group_pk,
    })
}

File: frost-taproot-go/group/group.go
// Package group provides high-level share set creation (trusted dealer).
package group

import (
	"github.com/frost-taproot/frost-taproot-go/shares"
	"github.com/frost-taproot/frost-taproot-go/types"
	"github.com/frost-taproot/frost-taproot-go/vss"
)

// CreateShareSet creates a set of secret shares and VSS commitments.
func CreateShareSet(threshold int, shareMax uint32, secrets [][32]byte) (types.SecretShareSet, error) {
	coeffs := vss.CreateShareCoeffs(secrets, threshold)
	sharesList, err := shares.CreateShares(coeffs, shareMax)
	if err != nil {
		return types.SecretShareSet{}, err
	}
	vssCommits := vss.GetShareCommits(coeffs)
	return types.SecretShareSet{
		Shares:     sharesList,
		VssCommits: vssCommits,
	}, nil
}

// CreateDealerSet creates a dealer share set with group public key.
func CreateDealerSet(threshold int, shareMax uint32, secrets [][32]byte) (types.DealerShareSet, error) {
	shareSet, err := CreateShareSet(threshold, shareMax, secrets)
	if err != nil {
		return types.DealerShareSet{}, err
	}
	return types.DealerShareSet{
		Shares:     shareSet.Shares,
		VssCommits: shareSet.VssCommits,
		GroupPk:    shareSet.VssCommits[0],
	}, nil
}
